jax/api.py

```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
+  convenience wrapper does not provide a mechanism for depending on anything
+  other than the function arguments, so JVP rules defined in this way can't
+  share work with the primal computation or with each other, though those things
+  are possible using ``jax.defjvp_all``. See ``jax.defjvp2`` for a variant that
+  allows for defining rules that depend on the output primal value. See also the
+  ``jax.defjvp_all`` docstring for more detail.
+
+  The signature of each component JVP rule is ``lambda g, *primals: ...`` where
+  ``g`` represents the tangent of the corresponding positional argument and
+  ``*primals`` represents all the primal positional arguments.
+
+  Defining a custom JVP rule also affects the default VJP rule, which is derived
+  from the JVP rule automatically via transposition.
+
+  Args:
+    fun: a custom_transforms function.
+    *jvprules: a sequence of functions or Nones specifying the JVP rule for each
+      corresponding positional argument. When an element is None, it indicates
+      that the Jacobian from the corresponding input to the output is zero.
+
+  Returns:
+    None. A side-effect is that ``fun`` is associated with the JVP rule
+    specified by ``*jvprules``.
+
+  For example:
+
+  >>> @jax.custom_transforms
+  ... def f(x):
+  ...   return np.sin(x ** 2)
+  ...
+  >>> print(f(3.))
+  0.4121185
+  >>> out_primal, out_tangent = jax.jvp(f, (3.,), (2.,))
+  >>> print(out_primal)
+  0.4121185
+  >>> print(out_tangent)
+  -10.933563
+  >>> jax.defjvp(f, lambda g, x: 8. * g)
+  >>> out_primal, out_tangent = jax.jvp(f, (3.,), (2.,))
+  >>> print(out_primal)
+  0.4121185
+  >>> print(out_tangent)
+  16.0
+
+  An example with a function on two arguments:
+
+  >>> @jax.custom_transforms
+  ... def f(x, y):
+  ...   return np.sin(x ** 2 + y)
+  ...
+  >>> print(f(3., 4.))
+  0.42016703
+  >>> out_primal, out_tangent = jax.jvp(f, (3., 4.), (1., 2.))
+  >>> print(out_primal)
+  0.42016703
+  >>> print(out_tangent)
+  7.2595744
+  >>> jax.defjvp(f, None, lambda g, x, y: 8. * g + y)
+  >>> out_primal, out_tangent = jax.jvp(f, (3., 4.), (1., 2.))
+  >>> print(out_primal)
+  0.42016703
+  >>> print(out_tangent)
+  20.0
+  """
+  _check_custom_transforms_type("defjvp", fun)
+  def custom_jvp(primals, tangents):
+    ans = fun(*primals)
+    tangents_out = [rule(t, *primals) for rule, t in zip(jvprules, tangents)
+                    if rule is not None and t is not ad_util.zero]
+    return ans, reduce(ad.add_tangents, tangents_out, ad_util.zero)
+  defjvp_all(fun, custom_jvp)
+
+def defjvp2(fun, *jvprules):
+  """Definine JVP rules for each argument separately.
+
+  This function is a convenience wrapper around ``jax.defjvp_all`` for separately
+  defining JVP rules for each of the function's arguments. This convenience
+  wrapper does not provide a mechanism for depending on anything other than the
+  function arguments and its primal output value.
```

Comment:
Maybe mention here or elsewhere that it's like defjvp, except the jvp function also has access the primal output.

Reply:
Done.